src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+                fds.poll_fds.swap_remove(data.poll_fds_index);
+                if let Some(swapped_pollfd) = fds.poll_fds.get(data.poll_fds_index) {
+                    fds.fd_data
+                        .get_mut(&swapped_pollfd.0.fd)
+                        .unwrap()
+                        .poll_fds_index = data.poll_fds_index;
+                }
+            }
+
+            Ok(())
+        })
+    }
+
+    pub fn wake(&self, token: Token) -> io::Result<()> {
+        let mut pending_wake_tokens = self.pending_wake_tokens.lock().unwrap();
+        pending_wake_tokens.push(token);
+        drop(pending_wake_tokens);
+        self.notify_waker.wake()
+    }
+}
+
+/// Shared record between IoSourceState and SelectorState that allows us to internally
+/// deregister partially or fully closed fds (i.e. when we get POLLHUP or PULLERR) without
+/// confusing IoSourceState and trying to deregister twice.  This isn't strictly
+/// required as technically deregister is idempotent but it is confusing
+/// when trying to debug behaviour as we get imbalanced calls to register/deregister and
+/// superfluous NotFound errors.
+#[derive(Debug)]
+pub(crate) struct RegistrationRecord {
+    is_unregistered: AtomicBool,
+}
+
+impl RegistrationRecord {
+    pub fn new() -> Self {
+        Self {
+            is_unregistered: AtomicBool::new(false),
+        }
+    }
+
+    pub fn mark_unregistered(&self) {
+        self.is_unregistered.store(true, Ordering::Relaxed);
+    }
+
+    #[allow(dead_code)]
+    pub fn is_registered(&self) -> bool {
+        !self.is_unregistered.load(Ordering::Relaxed)
+    }
+}
+
+#[cfg(target_os = "linux")]
+const POLLRDHUP: libc::c_short = libc::POLLRDHUP;
+#[cfg(not(target_os = "linux"))]
+const POLLRDHUP: libc::c_short = 0;
+
+const READ_EVENTS: libc::c_short = libc::POLLIN | POLLRDHUP;
+
+const WRITE_EVENTS: libc::c_short = libc::POLLOUT;
+
+const PRIORITY_EVENTS: libc::c_short = libc::POLLPRI;
+
+/// Get the input poll events for the given event.
+fn interests_to_poll(interest: Interest) -> libc::c_short {
+    let mut kind = 0;
+
+    if interest.is_readable() {
+        kind |= READ_EVENTS;
+    }
+
+    if interest.is_writable() {
+        kind |= WRITE_EVENTS;
+    }
+
+    if interest.is_priority() {
+        kind |= PRIORITY_EVENTS;
+    }
+
+    kind
+}
+
+/// Helper function to call poll.
+fn poll(fds: &mut [PollFd], deadline: Option<Instant>) -> io::Result<usize> {
+    loop {
+        // Convert the timeout to milliseconds.
+        let timeout_ms = deadline
+            .map(|deadline| {
+                let timeout = deadline.saturating_duration_since(Instant::now());
+
+                // Round up to a whole millisecond.
+                let mut ms = timeout.as_millis().try_into().unwrap_or(u64::MAX);
+                if Duration::from_millis(ms) < timeout {
+                    ms = ms.saturating_add(1);
+                }
+                ms.try_into().unwrap_or(i32::MAX)
```

Comment:
Can you use the same code as we use for the `epoll(2)` implementation: https://github.com/tokio-rs/mio/blob/2123bedbed00e24dccc036a0e1799e9e1d7dba1c/src/sys/unix/selector/epoll.rs#L90-L98

This way we have consistent timeout rounding.